ENH: Implement take_along_axis as described in #8708

eric-wieser opened the pull request:
Edit: Superceded by the simpler #11105 


See #8708 and earlier issues linked there for discussion of the need for this function.

Let's keep discussion here to the implementation.

In future, it would be nice to implement this with npyiter in C code for speed, but this is a good starting point, and likely just as fast as what is currently being used in the wild.

eric-wieser committed:
ENH: Add (put|take)_along_axis as described in #8708

This is the reduced version that does not allow any insertion of extra dimensions

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -104,13 +110,14 @@ def take(a, indices, axis=None, out=None, mode='raise'):
 
     Returns
     -------
-    subarray : ndarray
```
subarray was a bad word, because the result could be bigger than the input

mhvk commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,112 @@
+    array([30, 60])
+    >>> ai = np.argmax(a, axis=1); ai
+    array([1, 0], dtype=int64)
+    >>> np.take_along_axis(a, ai, axis=1)
```
Could you add an example where one keeps the dimension?

eric-wieser replied:
Again, I think keeping dimensions is out of scope here, and belongs in #8710

seberg replied:
In principle we could think about making keepdims (well, kind of the inverse) a kwarg here too.

seberg replied:
So you are lazy about a C version, too bad ;P

eric-wieser replied:
@seberg: Figured I'd implement it in python now, so that the tests can be set up, and there's a baseline for profiling

seberg replied:
Well, do the expand dims kind of thing I guess. Maybe its too much magic and users should just be forced to do keepdims (or feel the pain), keepdims is nice logic anyway.

eric-wieser replied:
I think it's too much magic, and doesn't even have a sensible meaning when `arr.ndim < ind.ndim`, which this also supports

eric-wieser replied:
What @mhvk is proposing, I think, is that:

```python
i_min = np.argmin(a, axis=1)
take_along_axis(a, i_min, axis=1, keepdims=1)
```
should mean either of (they are the same):
```python
i_min = np.argmin(a, axis=1)
take_along_axis(a, np.expand_dims(i_min, axis=1), axis=1)
```
```python
i_min = np.argmin(a, axis=1)
np.expand_dims( take_along_axis(a, i_min, axis=1), axis=1)
```


However, I think this is silly, and really we want to be able to write this
```python
i_min = np.argmin(a, axis=1, keepdims=True) # this is the line that stole our dim in the first place!
take_along_axis(a, i_min, axis=1)
```

seberg replied:
well, I mean you could implement `np.take_along_axis(a, indices, axis=1, expanddims=True)` in principle, but I am not convinced myself. I think this could be part of a mailing list discussion later.

seberg replied:
Do we want to support that case with the smaller dimensions?

eric-wieser replied:
@seberg: can you elaborate on which case you mean?

eric-wieser replied:
`arr.ndim < ind.ndim`?

That's for emulating the behaviour of `take` when passed a 2-or-more-d array.

Ignoring weird values of `axis`, `np.take(a, inds, axis)` is the same as:
```python
inds = inds.reshape((1,)*axis + inds.shape + (1,)*(a.ndim - axis - 1))
take_along_axis(a, inds, axis)
```

seberg replied:
@eric-wieser I did not follow your example fully yet. But I am not sure we should argue with take much. And broadcasting (and broadcasting alignment) does not seem trivial to give useful defaults for? Without a usecase or thinking about it much more, I would be -0.5 on allowing nifty stuff.

seberg replied:
IN some sense we can say that this `take` function is simply a vectorized version of take, the question is which dimensions take part in the indexing and which do not. The trivial case is "all take part". The indexing case would be "the first dimension", broadcasting might suggest "the last dimensions" (I am not sure what your take example would actually do).

eric-wieser replied:
> I am not sure what your take example would actually do

Which example are you talking about? All the examples in the docstring already contain the corresponding outptu

eric-wieser replied:
> Without a usecase or thinking about it much more, I would be -0.5 on allowing nifty stuff.

I guess my argument here is the zero-one-infinity rule. We need to support indexing with zero and 1d arrays for `argmax` and `argsort` respectively, so we should just support all dimensions of arrays. One (contrived) example of supporting this would be:

```
x = np.rand(2, 3, 4)
axis = 1
xi1 = np.argsort(x, axis=axis )
xi2 = np.argsort(abs(x), axis=axis )

xis = np.stack([xi1, xi2], axis=1)
x_and_ax_sorted = np.take_along_axis(x, xis, axis=1)
# x_and_ax_sorted[i,0,:,j] = sorted(x[i,:,j])
# x_and_ax_sorted[i,1,:,j] = sorted(x[i,:,j], key=abs)
```

Without this feature, you'd need to call `take_along_axis` twice if you have two different `argsort` results.


In the same ways as you can call `expand_dims` before or after `take_along_axis`, supporting arbitrary dimensionality allows you to call `stack` before or after `take_along_axis`

seberg replied:
Hmmm, first thought I misunderstood you, but yeah. So the thing is, why is it ` x_and_ax_sorted[i, 0, :, j]` and not e.g. `x_and_ax_sorted[0, i, :, j]` (or place the 0 to the end, of course with different stacking before)? It seems nontrivial to me what "is right", so I think we either skip it, or force the user to provide it maybe?

seberg replied:
Unless you actually meant more like `x_and_ax_soted[i, :x.shape[1], j]`

eric-wieser replied:
The rule here is that the middle axes of `indices` are inserted in place of the axis being indexed over. This is consistent with how `apply_along_axis` behaves as of #8441

eric-wieser replied:
> why is it 

Because the rule is that `x_and_ax_sorted[i,...,j] = x[i, xis[...], j]`

seberg replied:
And if the extra dimension in `xis` has the same size as th eother one, how do oyu know which one is which? Maybe I am just daft and should sleep and look at it more closely, but there feels like a can of disambiguities?

eric-wieser replied:
> how do oyu know which one is which

You know that all the axes that aren't the first `axis` axes or the last `ndim - axis - 1` axes are inserted, and copy these across to the output

seberg replied:
ok, nvm, I think I was just a bit daft

seberg replied:
I admit it is a safe and sensible generalization (similar to indexing), I am still not sure I like it too much.

eric-wieser replied:
>  I am still not sure I like it too much.

In the same way that I think `arr[:, 2dind, :]` is pretty wierd :p

seberg replied:
I guess the question is: Is it easy enough for any normal human to find? May it hide bugs? Is there maybe a better way to allow the thing without confusion? Maybe I am just a bit negative right now, but would prefer not to rush features that may be funky.

In principle one could define 4 types of axis: The axis along which we take, untouched axes, and "broadcast" axes iterated with the indexing array. The indexing array would have a matching amount of "broadcast" axis (plus "new" axes).

eric-wieser replied:
I count three axis types in that list. In my mind, the axes are:

 * The axis that is being removed in `arr`
 * All the remaining axes in `arr`, which are broadcast against the first and last axes of `ind`
 * The new axes being inserted in `ind`

`np.take` does not have this middle set of axes, but is otherwise the same.

I've updated the docstrings to try and better explain them, which I'll repeat here:

```python
out[i..., ..., k...] = a[i..., :, k...][indices]  # np.take
out[i..., ..., k...] = a[i..., :, k...][indices[i..., ..., k...]]  # np.take_along_axis
```

eric-wieser replied:
Are "subarray" axes just the ones I describe in my second bullet point?

seberg replied:
Well, yeah you can argue it is the same. But assuming the axis remapping for the broadcast may leave axes in the original array with no corresponding one in the indexing array. So yes, it is included, but only by adding broadcast dimensions to the indices.

eric-wieser replied:
Yeah, I guess my point is that `take_along_axis` is sufficiently general that you can build other operations (like `np.take`) out of it

seberg replied:
And I am philosophising what is actually useful ;), and if we can package *that* in a way that it is as easy as possible to figure out. Of course we can do your generalizations, but if nobody uses it anyway those generalizations may just end up being traps....

eric-wieser replied:
Would you prefer that for now, an `allow_dim_increase=False` flag is added that causes an error by default?

mhvk replied:
Just pitching in here: I do like the analogy with `apply_along_axis` and think we should not add another flag, but just keep it as is.

seberg replied:
Yes, but thinking about it a bit more, the analogy to `apply_along_axis` can also argue against the current proposal. Because if you would generalize apply along axis for multiple arrays, you would expect this:
```
apply_along_axis(axis, a, b)
a, b = broadcast(a, b)
for a_slice, b_slice in iterate_without_axis(a, b, axis=ax):
    func1d(a_slice, b_slice)
```
OTOH: Lets say we implement `np.searchsorted(arr, search, axis=1)` (or np.percentile which I think also only supports 1-d currently?), then implementing it to insert axes may make sense, and then making this the inverse would make sense too.

I repeat myself, but as long as I am not convinced that there is the "one obvious" way to define it, I am reluctant to freeze ourselves into a feature that in the end might not be compatible to e.g. the searchsorted generalization.

eric-wieser replied:
It seems to me that a `np.searchsorted` generalization would need to choose an answer the same decision that this does. For instance, I can see it being desireable to do

```python
out[i...,j,...,k...] = np.search_sorted_1d(arr[i..., :, j...], to_insert[i..., j..., k...])
```

eric-wieser replied:
I think I'm warming to actually only defining this for 1d indices, with the consequence that `take_alongaxis(a, a.argmax())` would not work directly, and would need an `expand_dims`

mhvk replied:
I feel that somewhat beats the point of this routine! Indeed, perhaps this is the strongest argument that your approach was the correct one: it works with all existing `arg*` methods (in their default form), and maybe future ones will have to adapt to work the same way.

eric-wieser replied:
I think the problem here comes down to where extra `(1,)` dimensions should be inserted, and I'm inclined to agree with @seberg that there are too many options, and forcing the user to specify is best.

So the rule would become

```python
assert(ind.ndim == x.ndim)
out[i..., :, j...] = x[i..., :, j...][ind[i..., :, j...]
```

Which would mean that you would write (`taa = take_along_axis`):

```python
# if you really want to squeeze that axis
max = taa(a, np.expand_dims(a.argmax(axis=axis), axis=axis), axis=axis).squeeze(axis=axis)
assert_equal(max, a.max(axis=axis))

# but lets be honest, squeezing is a bad plan for broadcasting
maxk = taa(a, a.argmax(axis=axis, keep_dims=True), axis=axis)
assert_equal(max, a.max(axis=axis, keep_dims=True))

# sort just works
sort = taa(a, a.argsort(axis=axis), axis=axis)
assert_equal(sort, a.sort(axis=axis, True))

# if you want to add another dimension, you make yourself explicit about where you want it
a_ext = np.expand_dims(a, axis=-1)
ai1 = np.argsort(a_ext, axis=axis )
ai2 = np.argsort(abs(a_ext), axis=axis )
ai = np.concatenate((ai1, ai2), axis=-1)
sort_a_abs_a = taa(a_ext, ai, axis=axis)

assert_equal(sort_a_abs_a[...,0], a_ext.sort(axis=axis))
assert_equal(sort_a_abs_a[...,1], abs(a_ext).sort(axis=axis))
```

seberg replied:
But there is no problem with argmax needing an expand dims, since we should just add a keepdims?

eric-wieser replied:
Right - it's inconvenient right now to have to call `np.expand_dims`, but that's really a bug elsewhere. And as for calling squeeze on the return value (vs what I proposed before) - you're likely in for a bad time if you start squeezing axes in the middle of your multidimensional data anyway.

Having `assert(ind.ndim == x.ndim)` still gives us plenty of freedom to come up with better semantics in future

eric-wieser replied:
Another example - indexing the brightest pixel:

```
assert img.shape == (400, 300, 3)
brightness = np.sum(img, axis=2, keepdims=True)  # again - always keep your dims!
argbrightrow = np.argmax(brightness, axis=1, keepdims=True)
brightest_by_row = np.take_along_axis(img, argbrightrow, axis=1)
```

eric-wieser replied:
> Having assert(ind.ndim == x.ndim) still gives us plenty of freedom to come up with better semantics in future

We could remove this and add the default broadcasting behaviour, but this risks confusing users to try to do `np.take_along_axis(a, a.argmax(axis=1),axis=1)`, which would allow this to sometimes silently do the wrong thing

seberg replied:
So for now, I guess lets just say we don't put it in, and put it up for discussion on the mailing list. If anyone can present a good reason/usecase especially for the way you first described it, I am good with allowing it. At least the 0-D case (dim is missing), it seems rather intuitive after all....

mhvk commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,112 @@
+
+    >>> np.max(a, axis=1)
+    array([30, 60])
+    >>> ai = np.argmax(a, axis=1); ai
```
I would not use `;` in example code; just make another line...

eric-wieser replied:
Resolved

mhvk commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,112 @@
+    else:
+        axis = normalize_axis_index(axis, arr.ndim)
+    if not _nx.issubdtype(indices.dtype, _nx.integer):
+        raise IndexError('arrays used as indices must be of integer type')
```
This should be a `TypeError`, I think (at least, `[1, 2][1.]` raises `TypeError`).

eric-wieser replied:
Just copying `np.array([1, 2])[np.array(1.)]` here, which gives `IndexError`. This is just that error message, but without the bit about booleans.

seberg replied:
Bools are not considered ints here right?

eric-wieser replied:
Correct, bools are not ints. There's a test for this

mhvk replied:
OK, should have tried with arrays; not very logical but best to stick with numpy practice here.

mhvk commented on numpy/lib/tests/test_shape_base.py:
```diff
@@ -3,14 +3,72 @@
 
 
+class TestTakeAlongAxis(TestCase):
+    def test_argequivalent(self):
```
Just to be sure, also add the `keepdims` versions.

eric-wieser replied:
I'd argue for postponing that to #8710 as well, and then no special handling would be needed

eric-wieser replied:
Ok, I've added a test to verify that `expand_dims` works before and after, which is sort of the same thing

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,112 @@
+
+    This computes::
+
+        out[a..., k..., b...] = arr[a..., indices[a..., k..., b...], b...]
```
Here I'm using the convention that `a = 1:A`, ie the capital letters are the shape, and the lowercase ones indices with that shape

eric-wieser replied:
There's no way I can use `axis` directly here, short of some ascii art pointing to the middle index on the right

eric-wieser replied:
I could use `out[i..., j..., k...] = arr[i..., indices[i..., j..., k...], k...]`, and then `Ni`, `Nk`, `Nk` further down?

mhvk replied:
I like the `i,j,k`, `Ni,Nj,Nk`, or perhaps `i1, i2, i3`, `N1,N2,N3`.

eric-wieser replied:
I've added a fixup commit to apply this. I'll squash once everything else is approved

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,112 @@
 
+def take_along_axis(arr, indices, axis):
+    """
+    Take the elements described by `indices` along each 1-D slice of the given
```
The problem is I need to disambiguate this from `take`, which is

    Take elements from an array along an axis.

seberg replied:
Should go to the mailing list in any case, maybe take is not even the most natural word in the end, which would somewhat remove the problem ;). The take description kind of only works for 1-D, and in 1-D the two do the same thing, so its a bit of a twist :).

mhvk replied:
It doesn't have to be my example, but I do think it should be a single line.

eric-wieser replied:
You're not wrong - I'm just asking for help in coming up with an unambiguous description under that constraint :)

seberg replied:
The only other idea I have right now is to call it a vectorized take/pick (which bites with the vindex idea, but maybe that name is not great anyway, I think someone had suggest broadcasted index there too, which may be more logical anyway -- though that does not necessarily mean easier I guess, hehe).

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,112 @@
+
+    >>> np.max(a, axis=1)
+    array([30, 60])
+    >>> ai = np.argmax(a, axis=1); ai
```
This is pretty typical - grep for `>>> (\w+)\s*=.*;\s*\1`

charris replied:
Bad practice anyway.

eric-wieser commented:
Does this last commit belong in this PR?

mhvk commented:
I think this is very nice. Generally, I think it is better to keep one PR to one logical commit, so in that sense the `np.ma.median` one should not be here, but it is nice to see an immediate use here. (In other words, either way is fine to me.)

Agreed though that this should be passed by the mailing list.

eric-wieser commented:
Oh, and I should probably add a `put_along_axis` with similar semantics, to match the `np.take / np.put` pair

eric-wieser commented:
`put_along_axis` is now in too. One thing I noticed while doing that is that `np.put` is not a very good dual to `np.take`, as it has no axis argument (#8765), and has a peculiar "repeat as necessary" behaviour

homu commented:
:umbrella: The latest upstream changes (presumably #8795) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #8847) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #8886) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
Could the reviewers involved here either sign off on it, merge it, or make a complaint.

eric-wieser commented:
@charris: The ball is in my court, I think - there was indecision about how to deal with broadcasting, with the suggestion of me consulting the mailing list - I have not done so.

I think it would be easier to design / put forth a case for this when `argmin` and `argmax` acquire keepdims arguments - so perhaps we should punt it to 1.14

charris commented:
OK, I'll punt. Thanks for the update.

homu commented:
:umbrella: The latest upstream changes (presumably #9050) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
Still in abeyance. Should I punt this on to 1.15?

seberg commented:
There is no much need for a specific milestone is there? I don't remember this, but for broadcasting, possibly we could do a minimal thing first that can be generalized later if it is too tricky to decide?

charris commented:
Removed the milestone.

eric-wieser commented:
I've rebased this just to avoid bitrot down the line, and moved the DOC commit to a separate pr (#9946). If nothing else, getting the doc change to existing functions into 1.14 makes it easier to pitch the new feature for 1.15.

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,192 @@
+
+    Or equivalently (where `...` alone is the builtin `Ellipsis`):
+
+        out[i..., ..., k...] = arr[i..., :, k...][indices[i..., ..., k...]]
```
Based on the discussion #9946, I suppose this would be better described as

````
Ni, Nk = a.shape[:axis], a.shape[axis+1:]
for ii in ndindex(Ni):
    for kk in ndindex(Nj):
        out[ii + s_[...,] + kk] = a[ii + s_[:,] + kk][indices[ii + s_[...,] + kk]]
```

rzu512 commented:
Hi. I am quite new to `git`. How to see your implementation for `take_along axis`, after I `git clone` this repository?

I wanted to do similar thing for a tree reduction, and came up with an implementation for `take_along_axis`.

The shape of the result is same as shape of the index.

<details>
<summary>take_along_axis</summary>

```python
def take_along_axis(arr, ind, axis):
    """Apply an index along axis to an array.
    Parameters
    ----------
        arr : np.ndarray
        ind : np.ndarray
            Indexing of an array along an axis. This cannot be an int.
            For a 2D array arr and axis=1, this means: from each column of
            arr, select elements along the column by the indices at the
            corresponding column of ind. In other words, the ith column of
            the result is arr[ind[:, i], i].
        axis : int
    Returns
    -------
        result : np.ndarray
    Example
    -------
        >>> import numpy as np
        >>> arr = np.flip(np.arange(8).reshape(2, 4), axis=1)  # test array
        >>> arr
        array([[3, 2, 1, 0],
               [7, 6, 5, 4]])
        >>> ind = np.argsort(arr, axis=1)  # indexing along axis 1
        >>> result = take_along_axis(arr, ind, axis=1)
        >>> result  # show the flat indexing
        array([[0, 1, 2, 3],
               [4, 5, 6, 7]])
        >>> answer = np.sort(arr, axis=1)
        >>> np.all(result == answer)
        True
    """
    # Does not check if axis or the ind are legal
    shape = arr.shape
    before = reduce(mul, shape[:axis], 1)
    at = shape[axis]
    after = arr.size // at // before
    a = arr.reshape(before, at, after)
    idx = [
        np.arange(before).reshape(before, 1, 1),
        ind.reshape(before, -1, after),
        np.arange(after).reshape(1, 1, after),
    ]
    return a[idx].reshape(ind.shape)
```
</details>

eric-wieser commented:
You'll need to add *my fork* of the repository (https://github.com/eric-wieser/numpy.git) as a remote, then you should just be able to checkout the `take_along_axis`.

Is your goal to compare my implementation with yours?

rzu512 commented:
Thanks. I found [it](https://github.com/eric-wieser/numpy/blob/8b7c244ffef4153b821cae2934dcae63fa123120/numpy/lib/shape_base.py#L23).

I mostly want to learn by looking at another solution.

crusaderky commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,192 @@
+    # normalize inputs
+    arr = asanyarray(arr)
+    indices = asanyarray(indices)
+    if axis is None:
```
Could you add support for arr.ndim > indices.ndim?
It's a single line here (+ unit test + docs):
```
indices = indices.reshape((1, ) * (arr.ndim - indices.ndim) + indices.shape)
```

Use case / example:
```
x = np.array([[5, 3, 2, 8, 1],
              [0, 7, 1, 3, 2]])
# Completely arbitrary y = f(x0, x1, ..., xn), embarassingly parallel along axis=-1
# Here we only have x0, but we could have more.
y = x.sum(axis=0)
# Sort the x's, moving the ones that cause the smallest y's to the left
take_along_axis(x, np.argsort(y))

eric-wieser replied:
I don't think your use-case is well-motivated. A more explicit way to achieve that would be:
```
y = x.sum(axis=0, keepdims=True)
take_along_axis(x, np.argsort(y, axis=1), axis=1)
```

crusaderky replied:
The whole point is that with the one-liner addition f(x) can add or remove axes at will (some manual broadcasting required if it _replaces_ axes, which however happens automatically if you e.g. wrap this in ``xarray.apply_ufunc``).

eric-wieser replied:
Note that in the comments above, we decide that perhaps it's best to not allow any case other than `indices.ndim == arr.ndim`, since there's no obvious right choice.

`take_along_axis` only really makes sense if you endeavor to keep all your axes aligned. `xarray` can probably solve that by axis names alone, but in numpy you need to indicate that by axis position. Therefore, you can't afford to let your axes collapse, and have numpy guess which one you lost: in your case, you're advocating for it to guess the left-most one should be reinserted - but this is only the case because you did `sum(axis=0)`.

crusaderky commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,192 @@
+    return tuple(fancy_index)
+
+
+def take_along_axis(arr, indices, axis):
```
default to axis=-1 like 99% of other numpy functions?

crusaderky commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,192 @@
+    return arr[_make_along_axis_idx(arr, indices, axis)]
+
+
+def put_along_axis(arr, indices, values, axis):
```
default to axis=-1 like 99% of other numpy functions?

eric-wieser replied:
I wish that were true. `concatenate` defaults to `axis=0`, and other functions default to `axis=None`

crusaderky reviewed (changes requested):
Waiting eagerly for this to get in! Any reason why there's no activity?

crusaderky commented on doc/release/1.14.0-notes.rst:
```diff
@@ -214,6 +214,20 @@ Chebyshev points of the first kind. A new ``Chebyshev.interpolate`` class
 method adds support for interpolation over arbitrary intervals using the scaled
 and shifted Chebyshev points of the first kind.
 
+New ``np.take_along_axis`` and ``np.put_along_axis`` functions
```
Needs moving to 1.15

crusaderky commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,192 @@
+
+        arr[i..., :, k...][indices[i..., ..., k...]] = values[i..., ..., k...]
+
+    .. versionadded:: 1.13.0
```
1.15.0

crusaderky commented on numpy/lib/shape_base.py:
```diff
@@ -16,9 +16,192 @@
+
+        out[i..., ..., k...] = arr[i..., :, k...][indices[i..., ..., k...]]
+
+    .. versionadded:: 1.13.0
```
1.15.0

eric-wieser committed:
ENH: Add expanded broadcasting rules to (take|put)_along_axis to match apply_along_axis

mhvk commented:
superceded by #11105, which was merged
